draft-ietf-tls-external-psk-importer.md

```diff
@@ -99,53 +101,68 @@ a tuple of (Base Key, External Identity, KDF). The associated KDF (and hash func
 - External Identity: The identity of an EPSK.
 - Imported Identity: The identity of a PSK as sent on the wire.
 
+## Notation and Cryptographic Dependencies
+
+The key importer API depends on a generic KDF which provides the following interface.
+
+- Nh: The output size of the Extract function.
+- Extract(salt, IKM): Extract a pseudorandom key of fixed length Nh
+  from input keying material `IKM` and an optional octet string
+  `salt`.
+- Expand(PRK, info, L): Expand a pseudorandom key `PRK` using
+  optional string `info` into `L` bytes of output keying material.
+
+It also depends on the following utility function.
+
+- `concat(x0, ..., xN)`: Concatenation of octet strings.
+  `concat(0x01, 0x0203, 0x040506) = 0x010203040506`
+
 # Key Import
 
 A key importer takes as input an EPSK with external identity 'external_identity' and base key 'epsk',
-as defined in {{terminology}}, along with an optional label, and transforms
-it into a set of PSKs and imported identities for use in a connection based on supported HashAlgorithms.
-In particular, for each supported HashAlgorithm 'hash', the importer constructs an ImportedIdentity
-structure as follows:
+as defined in {{terminology}}, along with an optional context string, and transforms
+it into a set of PSKs and imported identities for use in a connection based on the (D)TLS protocol
+version and a specific KDF. In particular, for each supported protocol version 'protocol' and KDF
+function 'kdf', the importer constructs an ImportedIdentity structure as follows:
 
 ~~~
    struct {
        opaque external_identity<1...2^16-1>;
-       opaque label<0..2^8-1>;
        opaque context<0..2^16-1>;
-       HashAlgorithm hash;
+       uint16 protocol;
+       uint16 kdf;
    } ImportedIdentity;
 ~~~
 
-[[TODO: An alternative design might combine label and hash into the same field so that future
-protocols which don't have a notion of HashAlgorithm don't need this field.]]
+ImportedIdentity.context MUST include the context used to derive the EPSK, if any exists.  If the EPSK is a key derived
```

Comment:
Is context something that you would use in derivation, or something you would do to further separate outputs based on context?  I tend to see this context as a way of taking an EPSK and applying it for a specific situation, not a way of providing information about how the EPSK originated (that would be baked into its identifier).

Say that I wanted to build a system that took Russ' pre-shared key + certificate idea.  I would have a group key agreement (or dissemination) protocol that produced a tuple of (key, identifier, KDF) which would form the EPSK.  In binding this into a TLS connection, I might want to take a hash of that protocol transcript into `context`, as you describe here.  But that doesn't really distinguish this usage from other potential uses in other protocols.  The target protocol and KDF help in the case that this is TLS vs. something completely different, but what if you want to use this for different TLS-using applications (HTTP and IMAP for example) and you want good key separation between those applications.  That's what I imagined this `context` field would be used for.

Maybe the right answer is to have multiple EPSK tuples produced from the source protocol, but that might not always be possible.

Either way, I think that you need to be clearer about whether `context` is for better distinguishing the input to this process, or in further distinguishing outputs.

Reply:
That sounds reasonable. I'll probably defer this to a separate cleanup issue for @jhoyla. He can better articulate the need for `context`. (In most cases, it will be empty, and I'll add that here.)

Reply:
The idea is that this says that this field lists everything that came before, i.e. the "history" of the KDF. The protocol field is for distinguishing the intended future use of the key, although maybe we need to make that more specific if you need to distinguish TLS for HTTP from TLS for IMAP.

If a commitment to future usage happens in a previous step that should appear in the transcript I guess? 

As a side note: If you're doing a group key agreement and then a PSK importer a simple transcript hash might not be sufficient, because the recipient doesn't know which of the group participants it is speaking to. You would need to include the client's group id also, and even then you need to trust group participants to be honest.